src/lib/server/lib.c

```diff
@@ -0,0 +1,230 @@
+/*
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
+ */
+
+/*
+ * $Id$
+ *
+ * @file lib.c
+ * @brief Handle global configuration, initialisation and freeing for libraries
+ *
+ * @copyright 2022 The FreeRADIUS server project
+ */
+#include <freeradius-devel/server/lib.h>
+#include <freeradius-devel/server/main_config.h>
+#include <freeradius-devel/util/atexit.h>
+
+/*
+ *  Terminator for array of fr_lib_autoinit_t
+ */
+fr_lib_autoinit_t const fr_lib_autoinit_terminator = { .name = NULL };
+
+/*
+ *  Global list of libraries
+ */
+static fr_lib_list_t *lib_list;
+
+/** Parse the global config section for a library and call its init function
+ *
+ * @param[in] lib	to configure and initialise
+ * @return
+ *	- 0 on success
+ *	- -1 on failure
+ */
+static int lib_init_call(fr_lib_t *lib)
+{
+	CONF_SECTION	*global_cs, *cs;
+
+	/*
+	 *	Find relevant config section
+	 */
+	global_cs = cf_section_find(cf_root(main_config->root_cs), "global", NULL);
+	if (!global_cs) {
+		WARN("Missing global config section");
+		goto init;
+	}
+
+	cs = cf_section_find(global_cs, lib->name, NULL);
+	if (!cs) {
+		WARN("Missing \"%s\" subsection in global config", lib->name);
+		goto init;
+	}
+
+	if ((cf_section_rules_push(cs, lib->autoinit->config)) < 0 ||
+	    (cf_section_parse(lib, lib->autoinit->inst, cs) < 0)) {
+		cf_log_err(cs, "Failed evaluating configuration for libldap");
+		return -1;
+	}
+
+	/*
+	 *  Call the init callback if defined
+	 */
+init:
+	if (lib->autoinit->init && (lib->autoinit->init()) < 0) return -1;
+
+	lib->initialised = true;
+
+	return 0;
+}
+
+/** Initialise a list of libraries
+ *
+ * @param to_init	Array of autoinit structures detailing libraries to initialise
+ * @return
+ *	- 0 on succcess
+ *	- -1 on failure
+ */
+static int lib_autoinit(fr_lib_autoinit_t * const *to_init)
+{
+	fr_lib_autoinit_t * const *p = to_init;
+
+	for (; (*p)->name; p++) {
```

Comment:
Or `p != fr_lib_autoinit_terminator`.  Ensures people actually use the terminator :)

Reply:
Maybe stick the init on the same line. Just makes it slightly clearer.